src/Dictionaries/RedisDictionarySource.cpp

```diff
@@ -250,23 +222,72 @@ namespace DB
                 if (isInteger(type))
                     key << DB::toString(key_columns[i]->get64(row));
                 else if (isString(type))
-                    key << get<String>((*key_columns[i])[row]);
+                    key << get<const String &>((*key_columns[i])[row]);
                 else
                     throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected type of key in Redis dictionary");
             }
 
             keys.add(key);
         }
 
-        return Pipe(std::make_shared<RedisSource>(client, std::move(keys), storage_type, sample_block, max_block_size));
+        return Pipe(std::make_shared<RedisSource>(std::move(conn), std::move(keys), storage_type, sample_block, REDIS_MAX_BLOCK_SIZE));
     }
 
-
     String RedisDictionarySource::toString() const
     {
         return "Redis: " + host + ':' + DB::toString(port);
     }
 
+    auto RedisDictionarySource::getConnection() const -> ConnectionPtr
+    {
+        ClientPtr client;
+        bool ok = pool->tryBorrowObject(client,
+            [] { return std::make_unique<Poco::Redis::Client>(); },
+            REDIS_LOCK_ACQUIRE_TIMEOUT_MS);
+
+        if (!ok)
+            throw Exception(ErrorCodes::TIMEOUT_EXCEEDED,
+                "Could not get connection from pool, timeout exceeded {} seconds",
+                REDIS_LOCK_ACQUIRE_TIMEOUT_MS);
+
+        if (!client->isConnected())
+        {
+            try
+            {
+                client->connect(host, port);
+
+                if (!password.empty())
```

Comment:
If client successfully connected, but this auth throws then, next time we get client from pool it will be connected and we will miss auth step.